lib/spack/spack/cmd/develop.py

```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
 
     spec = specs[0]
 
+    if args.recursive:
+        concrete_specs = env.all_matching_specs(spec)
+        if not concrete_specs:
+            tty.msg(
+                "No matching specs found in the environment. "
+                "Recursive develop requires a concretized environment"
+            )
+        else:
+            for s in concrete_specs:
+                for parent in s.traverse_edges(direction="parents", root=True):
+
+                    parent_args = copy(args)
+                    parent_args.spec = parent.spec.format("{name}@{version}")
```

Comment:
You want to use dag hashes... name@version is not specific enough.

Reply:
We can't use dag hashes b/c they will change when we concretize. I went generic to allow hitting multiple instances. I have always tried to go most generic for dev specs out of principle. I could do the `spec.short_spec` but if someone has a development environment with repeat specs (+cuda/~cuda) then `spack develop` can't support both.